Fix #18211: Enhance Validation for Video Start and End Time in Exploration Editor

joanapeixinho opened the pull request:
## Overview

1. This PR fixes https://github.com/oppia/oppia/issues/18211 
2. This PR does the following: Improved implementation for front-end validation has been integrated into the creator dashboard. This enhancement ensures that users can only submit videos if the start time precedes the end time; otherwise, an error prompt will be triggered the "Done" button is disabled. Additionally, it accommodates a special case where both start and end times are set to 0, enabling submission for playing the entire video. Furthermore, backend validation has been updated to maintain consistency across the platform. This PR is based on [another PR](https://github.com/oppia/oppia/pull/19179) made by @yungmulah.


## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [ ] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [ ] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)


## Proof that changes are correct


https://github.com/oppia/oppia/assets/99565826/31e1311e-fef0-4c3a-b0ff-3dbd97527b46

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

joanapeixinho committed:
fix oppia#18211: Unexpected error on Exploration Editor page RTE

joanapeixinho commented:
@seanlip @Lawful2002 PTAL

seanlip commented on core/templates/services/rte-helper-modal.controller.spec.ts:
```diff
@@ -108,9 +108,10 @@ describe('RteHelperModalComponent', () => {
-      component.attrsCustomizationArgsDict = {
-        heading: 'This value is not default.',
-      };
+      (component.componentId = 'video'),
```
Why not do these as two separate statements (separated by semicolon)?

joanapeixinho replied:
In fact I think it would be more readable if separated. I will change that!

seanlip commented on core/templates/services/rte-helper.service.spec.ts:
```diff
@@ -330,7 +330,6 @@ describe('Rte Helper Service', () => {
 
   it('should evalute when rich text component is not inline', () => {
     expect(rteHelperService.isInlineComponent('video')).toBe(false);
-    expect(rteHelperService.isInlineComponent('tabs')).toBe(false);
```
Why drop this line?

joanapeixinho replied:
I did it because on the other PR it was done but after looking into it I think it was erased by mistake since I can't see a reason to why it should be erased.

seanlip commented on core/templates/services/rte-helper.service.spec.ts:
```diff
@@ -330,7 +330,6 @@ describe('Rte Helper Service', () => {
 
   it('should evalute when rich text component is not inline', () => {
```
If possible, could you change evalute --> evaluate? Thanks!

seanlip reviewed (changes requested):
Thanks @joanapeixinho, it looks very nice!  Just a few small comments, and you might need to look into why the frontend CI checks are failing.

joanapeixinho committed:
Fix typo on create-activity-modal.component.spec.ts

joanapeixinho committed:
fix typo on rte-helper.service.spec.ts

joanapeixinho committed:
unsubscribe from customizationArgsFormSubscription in rte-helper-modal.controller.spec.ts

joanapeixinho committed:
Update initialization for test when customization args has a valid youtube video

joanapeixinho committed:
Fix indentation in beforeEach block

joanapeixinho commented:
@seanlip I made the changes and now I'm only failing one e2e test but I don't think I changed any code that could affect that test. Is that possible? Thank you for the help in advance :)

seanlip commented:
@joanapeixinho You'll probably need to look into it -- if you read the test logs, you'll see that they relate to the rich-text components, which is the part of the code you are modifying.

```
1) rich-text components should display rte involving file upload correctly
Error: Save State Content Button is not clickable.
```

joanapeixinho committed:
Merge branch 'oppia:develop' into fix-18211

joanapeixinho committed:
Refactor customizationArgsDict assignment in RteHelperModalComponent

joanapeixinho commented:
@seanlip I made some final changes, PTAL

seanlip commented on core/templates/services/rte-helper-modal.controller.ts:
```diff
@@ -218,7 +218,9 @@ export class RteHelperModalComponent {
     } else {
       this.ngbActiveModal.dismiss(false);
-      this.customizationArgsFormSubscription.unsubscribe();
+      if (this.customizationArgsFormSubscription) {
```
Could you explain why this change was needed (what's the context where the subscription is null)? Perhaps worth adding a comment.

joanapeixinho replied:
The test `describe('when customization args has a valid youtube video', function ()` in the file [rte-helper-modal.controller.spec.ts](https://github.com/oppia/oppia/blob/0d07f468562521cfd398a2bee0f407cd85ed9a18/core/templates/services/rte-helper-modal.controller.spec.ts) was failing with the error `Cannot read property 'unsubscribe' of undefined` so I thought that it was because of the case when components are opened, then canceled without suffering alterations so I added thar condition. 

Now, after taking a better look into it I think the test was failing because the componentIsNewlyCreated should be set as true in the test since that is the case and right now is just undefined. Can I change the test to set componentIsNewlyCreated to true?

joanapeixinho replied:
@seanlip Sorry, I did that because the test `should close modal when clicking on cancel button` was failing and I thought there was indeed a case where the subscription is null when we change the video link and then cancel right after. After looking into it I realized I was mistaken and the problem was that the test should call `component.ngOnInit();`. Can I change the test to call this function?

Also, I think in the cancel function the line  `this.customizationArgsFormSubscription.unsubscribe();` should be outside of the if clause since the subscription is created even when the component is newly created. 

Do these changes seem okay?

seanlip commented on core/templates/pages/creator-dashboard-page/modal-templates/create-activity-modal.component.spec.ts:
```diff
@@ -88,7 +88,7 @@ describe('Create Activity Modal Component', () => {
   });
 
-  it('should evalute component properties after component is initialized', fakeAsync(() => {
+  it('should evaluate component properties after component is initialized', fakeAsync(() => {
```
Thanks for fixing the additional places! :)

seanlip commented on core/templates/services/rte-helper-modal.controller.ts:
```diff
@@ -414,7 +416,14 @@ export class RteHelperModalComponent {
             this.tmpCustomizationArgs[i].value.toString()
           );
         }
+        (
```
I don't quite understand what's going on here ... can we find a way to avoid repeating the same logic in every branch of the code?

joanapeixinho replied:
I changed the save() function to be divided in one big if clause. Basically the process is very similar for all components that aren't mathExpression so I changed it to a way that I don't repeat the same logic. 

Done, PTAL at my new commit @seanlip

seanlip reviewed (changes requested):
Thanks @joanapeixinho! Just a note, make sure to reply to all comments (e.g. with "Done") -- see the PR instructions.

Left a couple more questions.

joanapeixinho committed:
Update save and cancel functions in RteHelperModalController

joanapeixinho committed:
Update test for when customization args has a valid youtube video

joanapeixinho committed:
Update save function rte-helper-modal.controller.ts

joanapeixinho committed:
Fix formating error in rte-helper-modal.controller.ts

joanapeixinho commented:
@seanlip I changed the code to match with your suggestions but I failed one end to end test that I believe as nothing to do with the part of the code I changed. Is that normal?

seanlip commented:
Hi @joanapeixinho, I think this one is a flake. I filed it at https://github.com/oppia/oppia/issues/20125 and restarted the test. Thanks!

seanlip commented on core/templates/services/rte-helper-modal.controller.ts:
```diff
@@ -382,29 +409,33 @@ export class RteHelperModalComponent {
+          customizationArgsDict as {
+            [Prop in CustomizationArgsNameAndValueArray[number]['name']]: CustomizationArgsNameAndValueArray[number]['value'];
+          }
+        )[caName] = this.tmpCustomizationArgs[i].value;
```
Wouldn't this have the effect of overwriting the 'text' customization arg in the link case that was only just set in lines 426-428?

joanapeixinho replied:
@seanlip The specific condition in lines 426-428  guarantees that the 'text' customization argument is never empty when it's overwritten by the general assignment. Therefore, the overwriting doesn't cause any issues since it's just set to its current value being the actuall text if not empty or the url if empty.

seanlip replied:
Hi @joanapeixinho, sorry for the late reply, I was extremely busy over the weekend.

I'm not sure I understand still, sorry. Suppose this.tmpCustomizationArgs[i - 1].value is nonempty and this.tmpCustomizationArgs[i].value is empty. Then lines 426-428 set the value of the arg to `this.tmpCustomizationArgs[i - 1].value`. But then this line, which runs after it, will set it back to the empty `this.tmpCustomizationArgs[i].value`. Right?

(Also if this analysis is correct, and the code is indeed wrong, then please write a test to catch this case in the future. Thanks!)

joanapeixinho replied:
@seanlip Done!

seanlip reviewed (changes requested):
Thanks @joanapeixinho! I think all the changes are fine except one, PTAL. If it is indeed an error, please add some frontend Karma tests to catch that case so that it doesn't break in the future.

seanlip commented:
Hi @joanapeixinho, friendly ping. Are you still looking at this?

joanapeixinho commented:
@seanlip Yes i'm still looking! Last week I had exams so I didn't have the time to look into this but I wanna work on it today and make a commit with the solution asap.

seanlip commented:
Thanks @joanapeixinho, sounds good. Please try to do it soon since I think there's another PR waiting on this one -- thanks!

seanlip commented:
Hi @joanapeixinho, friendly reminder on this, we will probably need to close this PR and open it up to the community again if the above isn't resolved in the next couple of days, since this blocks other fixes. Thanks!

joanapeixinho committed:
Update rte-helper-modal.controller.ts

joanapeixinho committed:
add test in rte-helper-modal.controller.spec.ts for when the text is empty in link form control

joanapeixinho commented:
@seanlip I made the changes you asked for, does it seem okay? Also I'm looking into why the e2e test failed.

HardikGoyal2003 commented:
Hey @joanapeixinho Please can you specify a date or a timeline on which you will be able to merge this PR, as its blocking one of my PRs. Thanks!

seanlip commented on core/templates/services/rte-helper-modal.controller.ts:
```diff
@@ -428,11 +428,13 @@ export class RteHelperModalComponent {
-            [Prop in CustomizationArgsNameAndValueArray[number]['name']]: CustomizationArgsNameAndValueArray[number]['value'];
-          }
-        )[caName] = this.tmpCustomizationArgs[i].value;
+        if (this.tmpCustomizationArgs[i].value) {
```
I'm not quite sure why you're introducing this "if" block ... wouldn't it be better to instead set this.tmpCustomizationArgs[i].value in line 422-428 instead of doing so for customizationArgsDics[caName] (similar to what you do for the video component) so that you can keep this part general.

Also reminder to please write some frontend tests for this, as requested previously, to ensure there aren't bugs in the logic.

joanapeixinho replied:
Done!

seanlip reviewed (changes requested):
Hi @joanapeixinho, left a comment, PTAL.  Thanks!

joanapeixinho committed:
Update rte-helper-modal.controller.ts

joanapeixinho commented:
@seanlip I made the changes and added a test for this situation. Thank u for the feedback! PTAL

seanlip reviewed (approved):
Thanks @joanapeixinho. LGTM!

seanlip commented:
@vojtechjelinek @nikitaevg PTAL as codeowners, thanks!

vojtechjelinek reviewed (approved):
LGTM!
